### PR TITLE
use VecGhostScatterBegin/VecGhostScatterEnd in IBFE interpolation

### PR DESCRIPTION
**Pull request by boyceg**

### CODE REVIEW

**Comment by boyceg** [comment]
Hmmm, what boneheaded git thing did I do to get 4f5fde3 showing up in this PR?

**Comment by boyceg** [comment]
I just reorganized the communication in `IBFEMethod::interpolateVelocity` to group together all of the calls to `VecGhostScatterBegin`, and all of the calls to `VecGhostScatterEnd`. I do not have time right now to test for correctness -- @drwells can you give this a try?

**Comment by drwells** [comment]
Nice. I'll give it a whirl.

**Comment by drwells** [comment]
This does seem to work, but I want to stare at it some more before approving it for merging.

**Comment by boyceg** [comment]
Does it make any performance difference?

**Comment by drwells** [comment]
I am working on that now: I am comparing it to the rest of the load balancing code.

**Comment by boyceg** [comment]
I just added some helper functions for batching calls to `VecCopy` and to `VecGhostUpdateBegin`/`VecGhostUpdateEnd`, and switched to using `VecGhostUpdateBegin`/`VecGhostUpdateEnd` in `IBFEMethod::spreadForce()`.

**Comment by boyceg** [comment]
Do you mean 6 procs or 16?

Note that this PR does a lot more index lookups to get the local indices for accumulating the ghost values — that could be slowing things down more than using the ghosted communication routines speeds things up. It should be fairly easy to cache this stuff if it is indeed the problem.

**Comment by boyceg** [comment]
Should we include some simple diagnostics to report min/mean/max cells/elems?

You may also need to play around with SAMRAI’s grid generation algorithm to get more reasonable looking patches.

**Comment by boyceg** [comment]
@knepley, there is an accumulation step that we could perform using the local form of a ghosted vector, or we can just use `VecSet`. (In `master` we are currently just using `VecSet` followed by `VecAssemblyXXX`. This PR switches to accumulating into the local form of the ghosted vectors followed by `VecGhostUpdateXXX`.) Do you have an opinion about whether we should be doing one vs. the other?

**Comment by boyceg** [comment]
This is superseded by #430.

**Comment by boyceg** [comment]
@drwells, this is the PR I was babbling about this afternoon. It uses the local form in `FEDataManager::interpWeighted()` to do the local part of the accumulation followed by a scatter to do parallel communication.

I will see if I can pull out the relevant stuff a bit later this evening.

File: ibtk/src/lagrangian/FEDataManager.cpp [hunk 1]
```diff
@@ -1387,6 +1436,17 @@ FEDataManager::interpWeighted(const int f_data_idx,
...
+                    // TODO: This probably should be cached.
+                    if (ghosted_vec)
+                    {
+                        F_local_dof_indices[i].resize(F_dof_indices[i].size());
+                        std::transform(F_dof_indices[i].begin(),
+                                       F_dof_indices[i].end(),
+                                       F_local_dof_indices[i].begin(),
+                                       [&](unsigned int idx) -> unsigned int {
+                                           return F_petsc_vec->map_global_to_local_index(idx);
+                                       });
```
Comment (drwells) [inline ibtk/src/lagrangian/FEDataManager.cpp]: This is, AFAIK, exactly what we need to do right now (until we get #513 finished)

File: ibtk/src/lagrangian/FEDataManager.cpp [hunk 2]
```diff
@@ -1270,17 +1302,34 @@ FEDataManager::interpWeighted(const int f_data_idx,
...
+        double* F_local_soln;
+        if (ghosted_vec)
+        {
+            ierr = VecGhostGetLocalForm(F_global_vec, &F_local_vec);
+            IBTK_CHKERRQ(ierr);
+            ierr = VecZeroEntries(F_local_vec);
+            IBTK_CHKERRQ(ierr);
+            ierr = VecGetArray(F_local_vec, &F_local_soln);
+            IBTK_CHKERRQ(ierr);
+        }
```
Comment (drwells) [inline ibtk/src/lagrangian/FEDataManager.cpp]: I think that this and the indexing are really all we need.

Comment (boyceg) [reply ibtk/src/lagrangian/FEDataManager.cpp]: I agree. I can do this later tonight (I think!).

Comment (drwells) [reply ibtk/src/lagrangian/FEDataManager.cpp]: @boyceg can you do this or shall I take a look?

**Comment by boyceg** [comment]
By all means --- I did not make it to this last night!

**Comment by drwells** [comment]
This should be a much smaller change now that the vectors are ghosted. Let me give it a spin.

**Pull request closed without merging**